// @swim/dataflow/lib/main/AbstractRecordOutlet.js
import { __extends } from "tslib";
import { Cursor } from "@swim/util";
import { BTree } from "@swim/collections";
import { Field, Record, Text } from "@swim/structure";
import { MapInlet, KeyOutlet } from "@swim/streamlet";
import { MemoizeMapCombinator } from "@swim/streamlet";
import { FilterFieldsCombinator } from "@swim/streamlet";
import { MapValueCombinator } from "@swim/streamlet";
import { MapFieldValuesCombinator } from "@swim/streamlet";
import { ReduceFieldsCombinator } from "@swim/streamlet";
import { WatchValueCombinator } from "@swim/streamlet";
import { WatchFieldsCombinator } from "@swim/streamlet";
import { RecordOutlet } from "./RecordOutlet";
var AbstractRecordOutlet = (function (_super) {
    __extends(AbstractRecordOutlet, _super);
    function AbstractRecordOutlet() {
        var _this = _super.call(this) || this;
        _this._effects = new BTree();
        _this._outlets = new BTree();
        _this._outputs = null;
        _this._version = -1;
        return _this;
    }
    AbstractRecordOutlet.prototype.streamletScope = function () {
        return null;
    };
    AbstractRecordOutlet.prototype.streamletContext = function () {
        var scope = this.streamletScope();
        if (scope !== null) {
            return scope.streamletContext();
        }
        return null;
    };
    AbstractRecordOutlet.prototype.hasOwn = function (key) {
        return this.has(key);
    };
    AbstractRecordOutlet.prototype.get = function (key) {
        if (key === void 0) {
            return this;
        }
        else {
            return _super.prototype.get.call(this, key);
        }
    };
    AbstractRecordOutlet.prototype.outlet = function (key) {
        if (typeof key === "string") {
            key = Text.from(key);
        }
        if (!this.hasOwn(key)) {
            var scope = this.streamletScope();
            if (RecordOutlet.is(scope) && scope.has(key)) {
                return scope.outlet(key);
            }
        }
        var outlet = this._outlets.get(key);
        if (outlet === void 0) {
            outlet = new KeyOutlet(this, key);
            this._outlets = this._outlets.updated(key, outlet);
            this.invalidateInputKey(key, 0);
        }
        return outlet;
    };
    AbstractRecordOutlet.prototype.outputIterator = function () {
        return this._outputs !== null ? Cursor.array(this._outputs) : Cursor.empty();
    };
    AbstractRecordOutlet.prototype.bindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        var newOutputs = new Array(n + 1);
        for (var i = 0; i < n; i += 1) {
            newOutputs[i] = oldOutputs[i];
        }
        newOutputs[n] = output;
        this._outputs = newOutputs;
    };
    AbstractRecordOutlet.prototype.unbindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        for (var i = 0; i < n; i += 1) {
            if (oldOutputs[i] === output) {
                if (n > 1) {
                    var newOutputs = new Array(n - 1);
                    for (var j = 0; j < i; j += 1) {
                        newOutputs[j] = oldOutputs[j];
                    }
                    for (var j = i; j < n - 1; j += 1) {
                        newOutputs[j] = oldOutputs[j + 1];
                    }
                    this._outputs = newOutputs;
                }
                else {
                    this._outputs = null;
                }
                break;
            }
        }
    };
    AbstractRecordOutlet.prototype.unbindOutputs = function () {
        var outlets = this._outlets;
        if (outlets.isEmpty()) {
            this._outlets = new BTree();
            outlets.forEach(function (key, keyOutlet) {
                keyOutlet.unbindOutputs();
            }, this);
        }
        var oldOutputs = this._outputs;
        if (oldOutputs !== null) {
            this._outputs = null;
            for (var i = 0, n = oldOutputs.length; i < n; i += 1) {
                oldOutputs[i].unbindInput();
            }
        }
    };
    AbstractRecordOutlet.prototype.disconnectOutputs = function () {
        var outlets = this._outlets;
        if (outlets.isEmpty()) {
            this._outlets = new BTree();
            outlets.forEach(function (key, keyOutlet) {
                keyOutlet.disconnectOutputs();
            }, this);
        }
        var outputs = this._outputs;
        if (outputs !== null) {
            this._outputs = null;
            for (var i = 0, n = outputs.length; i < n; i += 1) {
                var output = outputs[i];
                output.unbindInput();
                output.disconnectOutputs();
            }
        }
        this.forEach(function (member) {
            if (member instanceof Field) {
                member = member.toValue();
            }
            if (member instanceof AbstractRecordOutlet) {
                member.disconnectOutputs();
            }
            else if (member instanceof RecordOutlet.Streamlet) {
                member.disconnectOutputs();
            }
            else if (RecordOutlet.is(member)) {
                member.disconnectOutputs();
            }
        }, this);
    };
    AbstractRecordOutlet.prototype.disconnectInputs = function () {
    };
    AbstractRecordOutlet.prototype.invalidateInputKey = function (key, effect) {
        var oldEffects = this._effects;
        if (oldEffects.get(key) !== effect) {
            this.willInvalidateInputKey(key, effect);
            this._effects = oldEffects.updated(key, effect);
            this._version = -1;
            this.onInvalidateInputKey(key, effect);
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                var output = this._outputs[i];
                if (MapInlet.is(output)) {
                    output.invalidateOutputKey(key, effect);
                }
                else {
                    output.invalidateOutput();
                }
            }
            var outlet = this._outlets.get(key);
            if (outlet !== void 0) {
                outlet.invalidateInput();
            }
            this.didInvalidateInputKey(key, effect);
        }
    };
    AbstractRecordOutlet.prototype.invalidateInput = function () {
        if (this._version >= 0) {
            this.willInvalidateInput();
            this._version = -1;
            this.onInvalidateInput();
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                this._outputs[i].invalidateOutput();
            }
            this._outlets.forEach(function (key, outlet) {
                outlet.invalidateInput();
            }, this);
            this.didInvalidateInput();
        }
    };
    AbstractRecordOutlet.prototype.reconcileInputKey = function (key, version) {
        if (this._version < 0) {
            var oldEffects = this._effects;
            var effect = oldEffects.get(key);
            if (effect !== void 0) {
                this.willReconcileInputKey(key, effect, version);
                this._effects = oldEffects.removed(key);
                this.onReconcileInputKey(key, effect, version);
                for (var i = 0, n = this._outputs !== null ? this._outputs.length : 0; i < n; i += 1) {
                    var output = this._outputs[i];
                    if (MapInlet.is(output)) {
                        output.reconcileOutputKey(key, version);
                    }
                }
                var outlet = this._outlets.get(key);
                if (outlet !== void 0) {
                    outlet.reconcileInput(version);
                }
                this.didReconcileInputKey(key, effect, version);
            }
        }
    };
    AbstractRecordOutlet.prototype.reconcileInput = function (version) {
        if (this._version < 0) {
            this.willReconcileInput(version);
            this._effects.forEach(function (key) {
                this.reconcileInputKey(key, version);
            }, this);
            this._version = version;
            this.onReconcileInput(version);
            for (var i = 0, n = this._outputs !== null ? this._outputs.length : 0; i < n; i += 1) {
                this._outputs[i].reconcileOutput(version);
            }
            this.forEach(function (member) {
                if (member instanceof Field) {
                    member = member.toValue();
                }
                if (member instanceof AbstractRecordOutlet) {
                    member.reconcileInput(version);
                }
                else if (member instanceof RecordOutlet.Streamlet) {
                    member.reconcile(version);
                }
                else if (RecordOutlet.is(member)) {
                    member.reconcileInput(version);
                }
            }, this);
            this.didReconcileInput(version);
        }
    };
    AbstractRecordOutlet.prototype.willInvalidateInputKey = function (key, effect) {
    };
    AbstractRecordOutlet.prototype.onInvalidateInputKey = function (key, effect) {
    };
    AbstractRecordOutlet.prototype.didInvalidateInputKey = function (key, effect) {
    };
    AbstractRecordOutlet.prototype.willInvalidateInput = function () {
    };
    AbstractRecordOutlet.prototype.onInvalidateInput = function () {
    };
    AbstractRecordOutlet.prototype.didInvalidateInput = function () {
    };
    AbstractRecordOutlet.prototype.willReconcileInputKey = function (key, effect, version) {
    };
    AbstractRecordOutlet.prototype.onReconcileInputKey = function (key, effect, version) {
    };
    AbstractRecordOutlet.prototype.didReconcileInputKey = function (key, effect, version) {
    };
    AbstractRecordOutlet.prototype.willReconcileInput = function (version) {
    };
    AbstractRecordOutlet.prototype.onReconcileInput = function (version) {
    };
    AbstractRecordOutlet.prototype.didReconcileInput = function (version) {
    };
    AbstractRecordOutlet.prototype.memoize = function () {
        var combinator = new MemoizeMapCombinator();
        combinator.bindInput(this);
        return combinator;
    };
    AbstractRecordOutlet.prototype.filter = function (func) {
        if (typeof func !== "function") {
            return _super.prototype.filter.call(this, func);
        }
        else {
            var combinator = new FilterFieldsCombinator(func);
            combinator.bindInput(this);
            return combinator;
        }
    };
    AbstractRecordOutlet.prototype.map = function (func) {
        if (func.length === 1) {
            var combinator = new MapValueCombinator(func);
            combinator.bindInput(this);
            return combinator;
        }
        else {
            var combinator = new MapFieldValuesCombinator(func);
            combinator.bindInput(this);
            return combinator;
        }
    };
    AbstractRecordOutlet.prototype.reduce = function (identity, accumulator, combiner) {
        var combinator = new ReduceFieldsCombinator(identity, accumulator, combiner);
        combinator.bindInput(this);
        return combinator;
    };
    AbstractRecordOutlet.prototype.watch = function (func) {
        if (func.length === 1) {
            var combinator = new WatchValueCombinator(func);
            combinator.bindInput(this);
            return this;
        }
        else {
            var combinator = new WatchFieldsCombinator(func);
            combinator.bindInput(this);
            return this;
        }
    };
    return AbstractRecordOutlet;
}(Record));
export { AbstractRecordOutlet };
