// @swim/collections/lib/main/KeysCursor.js
import { __extends } from "tslib";
import { Cursor } from "@swim/util";
var KeysCursor = (function (_super) {
    __extends(KeysCursor, _super);
    function KeysCursor(cursor) {
        var _this = _super.call(this) || this;
        _this._cursor = cursor;
        return _this;
    }
    KeysCursor.prototype.isEmpty = function () {
        return this._cursor.isEmpty();
    };
    KeysCursor.prototype.head = function () {
        return this._cursor.head()[0];
    };
    KeysCursor.prototype.step = function () {
        this._cursor.step();
    };
    KeysCursor.prototype.skip = function (count) {
        this._cursor.skip(count);
    };
    KeysCursor.prototype.hasNext = function () {
        return this._cursor.hasNext();
    };
    KeysCursor.prototype.nextIndex = function () {
        return this._cursor.nextIndex();
    };
    KeysCursor.prototype.next = function () {
        var _a = this._cursor.next(), value = _a.value, done = _a.done;
        return { value: value && value[0], done: done };
    };
    KeysCursor.prototype.hasPrevious = function () {
        return this._cursor.hasPrevious();
    };
    KeysCursor.prototype.previousIndex = function () {
        return this._cursor.previousIndex();
    };
    KeysCursor.prototype.previous = function () {
        var _a = this._cursor.previous(), value = _a.value, done = _a.done;
        return { value: value && value[0], done: done };
    };
    KeysCursor.prototype.delete = function () {
        this._cursor.delete();
    };
    return KeysCursor;
}(Cursor));
export { KeysCursor };
