// @swim/collections/lib/main/ValuesCursor.js
import { __extends } from "tslib";
import { Cursor } from "@swim/util";
var ValuesCursor = (function (_super) {
    __extends(ValuesCursor, _super);
    function ValuesCursor(cursor) {
        var _this = _super.call(this) || this;
        _this._cursor = cursor;
        return _this;
    }
    ValuesCursor.prototype.isEmpty = function () {
        return this._cursor.isEmpty();
    };
    ValuesCursor.prototype.head = function () {
        return this._cursor.head()[1];
    };
    ValuesCursor.prototype.step = function () {
        this._cursor.step();
    };
    ValuesCursor.prototype.skip = function (count) {
        this._cursor.skip(count);
    };
    ValuesCursor.prototype.hasNext = function () {
        return this._cursor.hasNext();
    };
    ValuesCursor.prototype.nextIndex = function () {
        return this._cursor.nextIndex();
    };
    ValuesCursor.prototype.next = function () {
        var _a = this._cursor.next(), value = _a.value, done = _a.done;
        return { value: value && value[1], done: done };
    };
    ValuesCursor.prototype.hasPrevious = function () {
        return this._cursor.hasPrevious();
    };
    ValuesCursor.prototype.previousIndex = function () {
        return this._cursor.previousIndex();
    };
    ValuesCursor.prototype.previous = function () {
        var _a = this._cursor.previous(), value = _a.value, done = _a.done;
        return { value: value && value[1], done: done };
    };
    ValuesCursor.prototype.delete = function () {
        this._cursor.delete();
    };
    return ValuesCursor;
}(Cursor));
export { ValuesCursor };
