// @swim/uri/lib/main/UriHostIPv6.js
import { __extends } from "tslib";
import { Uri } from "./Uri";
import { UriHost } from "./UriHost";
var UriHostIPv6 = (function (_super) {
    __extends(UriHostIPv6, _super);
    function UriHostIPv6(address) {
        var _this = _super.call(this) || this;
        _this._address = address;
        return _this;
    }
    UriHostIPv6.prototype.address = function () {
        return this._address;
    };
    UriHostIPv6.prototype.ipv6 = function () {
        return this._address;
    };
    UriHostIPv6.prototype.debug = function (output) {
        output = output.write("UriHost").write(46).write("ipv6")
            .write(40).debug(this._address).write(41);
    };
    UriHostIPv6.prototype.display = function (output) {
        if (this._string != null) {
            output = output.write(this._string);
        }
        else {
            output = output.write(91);
            Uri.writeHostLiteral(this._address, output);
            output = output.write(93);
        }
    };
    UriHostIPv6.prototype.toString = function () {
        if (this._string == null) {
            this._string = "[" + this.address + "]";
        }
        return this._string;
    };
    return UriHostIPv6;
}(UriHost));
export { UriHostIPv6 };
Uri.HostIPv6 = UriHostIPv6;
