// @swim/uri/lib/main/UriPathSegment.js
import { __extends } from "tslib";
import { Format } from "@swim/codec";
import { Uri } from "./Uri";
import { UriPath } from "./UriPath";
var UriPathSegment = (function (_super) {
    __extends(UriPathSegment, _super);
    function UriPathSegment(head, tail) {
        var _this = _super.call(this) || this;
        _this._head = head;
        _this._tail = tail;
        return _this;
    }
    UriPathSegment.prototype.isDefined = function () {
        return true;
    };
    UriPathSegment.prototype.isAbsolute = function () {
        return false;
    };
    UriPathSegment.prototype.isRelative = function () {
        return true;
    };
    UriPathSegment.prototype.isEmpty = function () {
        return false;
    };
    UriPathSegment.prototype.head = function () {
        return this._head;
    };
    UriPathSegment.prototype.tail = function () {
        return this._tail;
    };
    UriPathSegment.prototype.setTail = function (tail) {
        if (tail.isAbsolute()) {
            this._tail = tail;
        }
        else {
            this._tail = UriPath.slash(tail);
        }
    };
    UriPathSegment.prototype.dealias = function () {
        return new UriPathSegment(this._head, this._tail);
    };
    UriPathSegment.prototype.parent = function () {
        var tail = this._tail;
        if (tail.isEmpty()) {
            return UriPath.empty();
        }
        else {
            var next = tail.tail();
            if (next.isEmpty()) {
                return UriPath.empty();
            }
            else {
                return new UriPathSegment(this._head, tail.parent());
            }
        }
    };
    UriPathSegment.prototype.base = function () {
        var tail = this._tail;
        if (tail.isEmpty()) {
            return UriPath.empty();
        }
        else {
            return new UriPathSegment(this._head, tail.base());
        }
    };
    UriPathSegment.prototype.prependedSegment = function (segment) {
        return UriPath.segment(segment, UriPath.slash(this));
    };
    UriPathSegment.prototype.debug = function (output) {
        output = output.write("UriPath").write(46).write("parse")
            .write(40).write(34).display(this).write(34).write(41);
    };
    UriPathSegment.prototype.display = function (output) {
        if (this._string !== void 0) {
            output = output.write(this._string);
        }
        else {
            _super.prototype.display.call(this, output);
        }
    };
    UriPathSegment.prototype.toString = function () {
        if (this._string === void 0) {
            this._string = Format.display(this);
        }
        return this._string;
    };
    return UriPathSegment;
}(UriPath));
export { UriPathSegment };
Uri.PathSegment = UriPathSegment;
