// @swim/structure/lib/main/TextOutput.js
import { __extends } from "tslib";
import { OutputSettings, Output } from "@swim/codec";
import { Text } from "./Text";
var TextOutput = (function (_super) {
    __extends(TextOutput, _super);
    function TextOutput(string, settings) {
        var _this = _super.call(this) || this;
        _this._string = string;
        _this._settings = settings;
        return _this;
    }
    TextOutput.prototype.isCont = function () {
        return true;
    };
    TextOutput.prototype.isFull = function () {
        return false;
    };
    TextOutput.prototype.isDone = function () {
        return false;
    };
    TextOutput.prototype.isError = function () {
        return false;
    };
    TextOutput.prototype.isPart = function (isPart) {
        if (isPart === void 0) {
            return false;
        }
        else {
            return this;
        }
    };
    TextOutput.prototype.write = function (token) {
        if (typeof token === "number") {
            if ((token >= 0x0000 && token <= 0xd7ff)
                || (token >= 0xe000 && token <= 0xffff)) {
                token = String.fromCharCode(token);
            }
            else if (token >= 0x10000 && token <= 0x10ffff) {
                var u = token - 0x10000;
                token = String.fromCharCode(0xd800 | (u >>> 10), 0xdc00 | (u & 0x3ff));
            }
            else {
                token = "\ufffd";
            }
        }
        this._string += token;
        return this;
    };
    TextOutput.prototype.settings = function (settings) {
        if (settings === void 0) {
            return this._settings;
        }
        else {
            this._settings = OutputSettings.fromAny(settings);
            return this;
        }
    };
    TextOutput.prototype.bind = function () {
        return Text.from(this._string);
    };
    TextOutput.prototype.clone = function () {
        return new TextOutput(this._string, this._settings);
    };
    TextOutput.prototype.toString = function () {
        return this._string;
    };
    return TextOutput;
}(Output));
export { TextOutput };
