// @swim/client/lib/main/ref/HostRef.js
import { __extends } from "tslib";
import { Uri } from "@swim/uri";
import { EventDownlink } from "../downlink/EventDownlink";
import { ListDownlink } from "../downlink/ListDownlink";
import { MapDownlink } from "../downlink/MapDownlink";
import { ValueDownlink } from "../downlink/ValueDownlink";
import { BaseRef } from "./BaseRef";
import { NodeRef } from "./NodeRef";
import { LaneRef } from "./LaneRef";
var HostRef = (function (_super) {
    __extends(HostRef, _super);
    function HostRef(context, hostUri) {
        var _this = _super.call(this, context) || this;
        _this._hostUri = hostUri;
        return _this;
    }
    HostRef.prototype.hostUri = function () {
        return this._hostUri;
    };
    HostRef.prototype.hostRef = function (hostUri) {
        hostUri = Uri.fromAny(hostUri);
        return new HostRef(this._context, hostUri);
    };
    HostRef.prototype.nodeRef = function (nodeUri) {
        nodeUri = Uri.fromAny(nodeUri);
        return new NodeRef(this._context, this._hostUri, nodeUri);
    };
    HostRef.prototype.laneRef = function (nodeUri, laneUri) {
        nodeUri = Uri.fromAny(nodeUri);
        laneUri = Uri.fromAny(laneUri);
        return new LaneRef(this._context, this._hostUri, nodeUri, laneUri);
    };
    HostRef.prototype.downlink = function (init) {
        return new EventDownlink(this._context, this, init, this._hostUri);
    };
    HostRef.prototype.downlinkList = function (init) {
        return new ListDownlink(this._context, this, init, this._hostUri);
    };
    HostRef.prototype.downlinkMap = function (init) {
        return new MapDownlink(this._context, this, init, this._hostUri);
    };
    HostRef.prototype.downlinkValue = function (init) {
        return new ValueDownlink(this._context, this, init, this._hostUri);
    };
    HostRef.prototype.command = function (nodeUri, laneUri, body) {
        this._context.command(this._hostUri, nodeUri, laneUri, body);
    };
    return HostRef;
}(BaseRef));
export { HostRef };
