// @swim/client/lib/main/ref/LaneRef.js
import { __extends } from "tslib";
import { Uri } from "@swim/uri";
import { EventDownlink } from "../downlink/EventDownlink";
import { ListDownlink } from "../downlink/ListDownlink";
import { MapDownlink } from "../downlink/MapDownlink";
import { ValueDownlink } from "../downlink/ValueDownlink";
import { BaseRef } from "./BaseRef";
import { HostRef } from "./HostRef";
import { NodeRef } from "./NodeRef";
var LaneRef = (function (_super) {
    __extends(LaneRef, _super);
    function LaneRef(context, hostUri, nodeUri, laneUri) {
        var _this = _super.call(this, context) || this;
        _this._hostUri = hostUri;
        _this._nodeUri = nodeUri;
        _this._laneUri = laneUri;
        return _this;
    }
    LaneRef.prototype.hostUri = function () {
        return this._hostUri;
    };
    LaneRef.prototype.nodeUri = function () {
        return this._nodeUri;
    };
    LaneRef.prototype.laneUri = function () {
        return this._laneUri;
    };
    LaneRef.prototype.hostRef = function (hostUri) {
        hostUri = Uri.fromAny(hostUri);
        return new HostRef(this._context, hostUri);
    };
    LaneRef.prototype.nodeRef = function (nodeUri) {
        nodeUri = Uri.fromAny(nodeUri);
        return new NodeRef(this._context, this._hostUri, nodeUri);
    };
    LaneRef.prototype.laneRef = function (laneUri) {
        laneUri = Uri.fromAny(laneUri);
        return new LaneRef(this._context, this._hostUri, this._nodeUri, laneUri);
    };
    LaneRef.prototype.downlink = function (init) {
        return new EventDownlink(this._context, this, init, this._hostUri, this._nodeUri, this._laneUri);
    };
    LaneRef.prototype.downlinkList = function (init) {
        return new ListDownlink(this._context, this, init, this._hostUri, this._nodeUri, this._laneUri);
    };
    LaneRef.prototype.downlinkMap = function (init) {
        return new MapDownlink(this._context, this, init, this._hostUri, this._nodeUri, this._laneUri);
    };
    LaneRef.prototype.downlinkValue = function (init) {
        return new ValueDownlink(this._context, this, init, this._hostUri, this._nodeUri, this._laneUri);
    };
    LaneRef.prototype.command = function (body) {
        this._context.command(this._hostUri, this._nodeUri, this._laneUri, body);
    };
    return LaneRef;
}(BaseRef));
export { LaneRef };
