// @swim/client/lib/main/ref/NodeRef.js
import { __extends } from "tslib";
import { Uri } from "@swim/uri";
import { EventDownlink } from "../downlink/EventDownlink";
import { ListDownlink } from "../downlink/ListDownlink";
import { MapDownlink } from "../downlink/MapDownlink";
import { ValueDownlink } from "../downlink/ValueDownlink";
import { BaseRef } from "./BaseRef";
import { HostRef } from "./HostRef";
import { LaneRef } from "./LaneRef";
var NodeRef = (function (_super) {
    __extends(NodeRef, _super);
    function NodeRef(context, hostUri, nodeUri) {
        var _this = _super.call(this, context) || this;
        _this._hostUri = hostUri;
        _this._nodeUri = nodeUri;
        return _this;
    }
    NodeRef.prototype.hostUri = function () {
        return this._hostUri;
    };
    NodeRef.prototype.nodeUri = function () {
        return this._nodeUri;
    };
    NodeRef.prototype.hostRef = function (hostUri) {
        hostUri = Uri.fromAny(hostUri);
        return new HostRef(this._context, hostUri);
    };
    NodeRef.prototype.nodeRef = function (nodeUri) {
        nodeUri = Uri.fromAny(nodeUri);
        return new NodeRef(this._context, this._hostUri, nodeUri);
    };
    NodeRef.prototype.laneRef = function (laneUri) {
        laneUri = Uri.fromAny(laneUri);
        return new LaneRef(this._context, this._hostUri, this._nodeUri, laneUri);
    };
    NodeRef.prototype.downlink = function (init) {
        return new EventDownlink(this._context, this, init, this._hostUri, this._nodeUri);
    };
    NodeRef.prototype.downlinkList = function (init) {
        return new ListDownlink(this._context, this, init, this._hostUri, this._nodeUri);
    };
    NodeRef.prototype.downlinkMap = function (init) {
        return new MapDownlink(this._context, this, init, this._hostUri, this._nodeUri);
    };
    NodeRef.prototype.downlinkValue = function (init) {
        return new ValueDownlink(this._context, this, init, this._hostUri, this._nodeUri);
    };
    NodeRef.prototype.command = function (laneUri, body) {
        this._context.command(this._hostUri, this._nodeUri, laneUri, body);
    };
    return NodeRef;
}(BaseRef));
export { NodeRef };
