// @swim/streamlet/lib/main/AbstractInoutlet.js
import { Cursor } from "@swim/util";
import { Outlet } from "./Outlet";
var AbstractInoutlet = (function () {
    function AbstractInoutlet() {
        this._input = null;
        this._outputs = null;
        this._version = -1;
    }
    AbstractInoutlet.prototype.input = function () {
        return this._input;
    };
    AbstractInoutlet.prototype.bindInput = function (input) {
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = input;
        if (this._input !== null) {
            this._input.bindOutput(this);
        }
    };
    AbstractInoutlet.prototype.unbindInput = function () {
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = null;
    };
    AbstractInoutlet.prototype.disconnectInputs = function () {
        if (this._outputs === null) {
            var input = this._input;
            if (input !== null) {
                input.unbindOutput(this);
                this._input = null;
                input.disconnectInputs();
            }
        }
    };
    AbstractInoutlet.prototype.outputIterator = function () {
        return this._outputs !== null ? Cursor.array(this._outputs) : Cursor.empty();
    };
    AbstractInoutlet.prototype.bindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        var newOutputs = new Array(n + 1);
        for (var i = 0; i < n; i += 1) {
            newOutputs[i] = oldOutputs[i];
        }
        newOutputs[n] = output;
        this._outputs = newOutputs;
    };
    AbstractInoutlet.prototype.unbindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        for (var i = 0; i < n; i += 1) {
            if (oldOutputs[i] === output) {
                if (n > 1) {
                    var newOutputs = new Array(n - 1);
                    for (var j = 0; j < i; j += 1) {
                        newOutputs[j] = oldOutputs[j];
                    }
                    for (var j = i; j < n - 1; j += 1) {
                        newOutputs[j] = oldOutputs[j + 1];
                    }
                    this._outputs = newOutputs;
                }
                else {
                    this._outputs = null;
                }
                break;
            }
        }
    };
    AbstractInoutlet.prototype.unbindOutputs = function () {
        var outputs = this._outputs;
        if (outputs !== null) {
            this._outputs = null;
            for (var i = 0, n = outputs.length; i < n; i += 1) {
                var output = outputs[i];
                output.unbindInput();
            }
        }
    };
    AbstractInoutlet.prototype.disconnectOutputs = function () {
        if (this._input === null) {
            var outputs = this._outputs;
            if (outputs !== null) {
                this._outputs = null;
                for (var i = 0, n = outputs.length; i < n; i += 1) {
                    var output = outputs[i];
                    output.unbindInput();
                    output.disconnectOutputs();
                }
            }
        }
    };
    AbstractInoutlet.prototype.invalidateOutput = function () {
        this.invalidate();
    };
    AbstractInoutlet.prototype.invalidateInput = function () {
        this.invalidate();
    };
    AbstractInoutlet.prototype.invalidate = function () {
        if (this._version >= 0) {
            this.willInvalidate();
            this._version = -1;
            this.onInvalidate();
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                this._outputs[i].invalidateOutput();
            }
            this.didInvalidate();
        }
    };
    AbstractInoutlet.prototype.reconcileOutput = function (version) {
        this.reconcile(version);
    };
    AbstractInoutlet.prototype.reconcileInput = function (version) {
        this.reconcile(version);
    };
    AbstractInoutlet.prototype.reconcile = function (version) {
        if (this._version < 0) {
            this.willReconcile(version);
            this._version = version;
            if (this._input !== null) {
                this._input.reconcileInput(version);
            }
            this.onReconcile(version);
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                this._outputs[i].reconcileOutput(version);
            }
            this.didReconcile(version);
        }
    };
    AbstractInoutlet.prototype.willInvalidate = function () {
    };
    AbstractInoutlet.prototype.onInvalidate = function () {
    };
    AbstractInoutlet.prototype.didInvalidate = function () {
    };
    AbstractInoutlet.prototype.willReconcile = function (version) {
    };
    AbstractInoutlet.prototype.onReconcile = function (version) {
    };
    AbstractInoutlet.prototype.didReconcile = function (version) {
    };
    AbstractInoutlet.prototype.memoize = function () {
        var combinator = new Outlet.MemoizeValueCombinator();
        combinator.bindInput(this);
        return combinator;
    };
    AbstractInoutlet.prototype.map = function (func) {
        var combinator = new Outlet.MapValueCombinator(func);
        combinator.bindInput(this);
        return combinator;
    };
    AbstractInoutlet.prototype.watch = function (func) {
        var combinator = new Outlet.WatchValueCombinator(func);
        combinator.bindInput(this);
        return this;
    };
    return AbstractInoutlet;
}());
export { AbstractInoutlet };
