// @swim/streamlet/lib/main/AbstractMapInoutlet.js
import { Cursor } from "@swim/util";
import { BTree } from "@swim/collections";
import { Outlet } from "./Outlet";
import { MapInlet } from "./MapInlet";
import { MapOutlet } from "./MapOutlet";
import { KeyOutlet } from "./KeyOutlet";
var AbstractMapInoutlet = (function () {
    function AbstractMapInoutlet() {
        this._input = null;
        this._effects = new BTree();
        this._outlets = new BTree();
        this._outputs = null;
        this._version = -1;
    }
    AbstractMapInoutlet.prototype.input = function () {
        return this._input;
    };
    AbstractMapInoutlet.prototype.bindInput = function (input) {
        if (!MapOutlet.is(input)) {
            throw new TypeError("" + input);
        }
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = input;
        if (this._input !== null) {
            this._input.bindOutput(this);
        }
    };
    AbstractMapInoutlet.prototype.unbindInput = function () {
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = null;
    };
    AbstractMapInoutlet.prototype.disconnectInputs = function () {
        if (this._outputs === null && this._outlets.isEmpty()) {
            var input = this._input;
            if (input !== null) {
                input.unbindOutput(this);
                this._input = null;
                input.disconnectInputs();
            }
        }
    };
    AbstractMapInoutlet.prototype.outlet = function (key) {
        var outlet = this._outlets.get(key);
        if (outlet === void 0) {
            outlet = new KeyOutlet(this, key);
            this._outlets = this._outlets.updated(key, outlet);
        }
        return outlet;
    };
    AbstractMapInoutlet.prototype.outputIterator = function () {
        return this._outputs !== null ? Cursor.array(this._outputs) : Cursor.empty();
    };
    AbstractMapInoutlet.prototype.bindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        var newOutputs = new Array(n + 1);
        for (var i = 0; i < n; i += 1) {
            newOutputs[i] = oldOutputs[i];
        }
        newOutputs[n] = output;
        this._outputs = newOutputs;
    };
    AbstractMapInoutlet.prototype.unbindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        for (var i = 0; i < n; i += 1) {
            if (oldOutputs[i] === output) {
                if (n > 1) {
                    var newOutputs = new Array(n - 1);
                    for (var j = 0; j < i; j += 1) {
                        newOutputs[j] = oldOutputs[j];
                    }
                    for (var j = i; j < n - 1; j += 1) {
                        newOutputs[j] = oldOutputs[j + 1];
                    }
                    this._outputs = newOutputs;
                }
                else {
                    this._outputs = null;
                }
                break;
            }
        }
    };
    AbstractMapInoutlet.prototype.unbindOutputs = function () {
        var outlets = this._outlets;
        if (outlets.isEmpty()) {
            this._outlets = new BTree();
            outlets.forEach(function (key, keyOutlet) {
                keyOutlet.unbindOutputs();
            }, this);
        }
        var outputs = this._outputs;
        if (outputs !== null) {
            this._outputs = null;
            for (var i = 0, n = outputs.length; i < n; i += 1) {
                var output = outputs[i];
                output.unbindInput();
            }
        }
    };
    AbstractMapInoutlet.prototype.disconnectOutputs = function () {
        if (this._input === null) {
            var outlets = this._outlets;
            if (outlets.isEmpty()) {
                this._outlets = new BTree();
                outlets.forEach(function (key, keyOutlet) {
                    keyOutlet.disconnectOutputs();
                }, this);
            }
            var outputs = this._outputs;
            if (outputs !== null) {
                this._outputs = null;
                for (var i = 0, n = outputs.length; i < n; i += 1) {
                    var output = outputs[i];
                    output.unbindInput();
                    output.disconnectOutputs();
                }
            }
        }
    };
    AbstractMapInoutlet.prototype.invalidateOutputKey = function (key, effect) {
        this.invalidateKey(key, effect);
    };
    AbstractMapInoutlet.prototype.invalidateInputKey = function (key, effect) {
        this.invalidateKey(key, effect);
    };
    AbstractMapInoutlet.prototype.invalidateKey = function (key, effect) {
        var oldEffects = this._effects;
        if (oldEffects.get(key) !== effect) {
            this.willInvalidateKey(key, effect);
            this._effects = oldEffects.updated(key, effect);
            this._version = -1;
            this.onInvalidateKey(key, effect);
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                var output = this._outputs[i];
                if (MapInlet.is(output)) {
                    output.invalidateOutputKey(key, effect);
                }
                else {
                    output.invalidateOutput();
                }
            }
            var outlet = this._outlets.get(key);
            if (outlet !== void 0) {
                outlet.invalidateInput();
            }
            this.didInvalidateKey(key, effect);
        }
    };
    AbstractMapInoutlet.prototype.invalidateOutput = function () {
        this.invalidate();
    };
    AbstractMapInoutlet.prototype.invalidateInput = function () {
        this.invalidate();
    };
    AbstractMapInoutlet.prototype.invalidate = function () {
        if (this._version >= 0) {
            this.willInvalidate();
            this._version = -1;
            this.onInvalidate();
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                this._outputs[i].invalidateOutput();
            }
            this._outlets.forEach(function (key, outlet) {
                outlet.invalidateInput();
            }, this);
            this.didInvalidate();
        }
    };
    AbstractMapInoutlet.prototype.reconcileOutputKey = function (key, version) {
        this.reconcileKey(key, version);
    };
    AbstractMapInoutlet.prototype.reconcileInputKey = function (key, version) {
        this.reconcileKey(key, version);
    };
    AbstractMapInoutlet.prototype.reconcileKey = function (key, version) {
        if (this._version < 0) {
            var oldEffects = this._effects;
            var effect = oldEffects.get(key);
            if (effect !== void 0) {
                this.willReconcileKey(key, effect, version);
                this._effects = oldEffects.removed(key);
                if (this._input !== null) {
                    this._input.reconcileInputKey(key, version);
                }
                this.onReconcileKey(key, effect, version);
                for (var i = 0, n = this._outputs !== null ? this._outputs.length : 0; i < n; i += 1) {
                    var output = this._outputs[i];
                    if (MapInlet.is(output)) {
                        output.reconcileOutputKey(key, version);
                    }
                }
                var outlet = this._outlets.get(key);
                if (outlet !== void 0) {
                    outlet.reconcileInput(version);
                }
                this.didReconcileKey(key, effect, version);
            }
        }
    };
    AbstractMapInoutlet.prototype.reconcileOutput = function (version) {
        this.reconcile(version);
    };
    AbstractMapInoutlet.prototype.reconcileInput = function (version) {
        this.reconcile(version);
    };
    AbstractMapInoutlet.prototype.reconcile = function (version) {
        if (this._version < 0) {
            this.willReconcile(version);
            this._effects.forEach(function (key) {
                this.reconcileKey(key, version);
            }, this);
            this._version = version;
            this.onReconcile(version);
            for (var i = 0, n = this._outputs !== null ? this._outputs.length : 0; i < n; i += 1) {
                this._outputs[i].reconcileOutput(version);
            }
            this.didReconcile(version);
        }
    };
    AbstractMapInoutlet.prototype.willInvalidateKey = function (key, effect) {
    };
    AbstractMapInoutlet.prototype.onInvalidateKey = function (key, effect) {
    };
    AbstractMapInoutlet.prototype.didInvalidateKey = function (key, effect) {
    };
    AbstractMapInoutlet.prototype.willInvalidate = function () {
    };
    AbstractMapInoutlet.prototype.onInvalidate = function () {
    };
    AbstractMapInoutlet.prototype.didInvalidate = function () {
    };
    AbstractMapInoutlet.prototype.willReconcileKey = function (key, effect, version) {
    };
    AbstractMapInoutlet.prototype.onReconcileKey = function (key, effect, version) {
    };
    AbstractMapInoutlet.prototype.didReconcileKey = function (key, effect, version) {
    };
    AbstractMapInoutlet.prototype.willReconcile = function (version) {
    };
    AbstractMapInoutlet.prototype.onReconcile = function (version) {
    };
    AbstractMapInoutlet.prototype.didReconcile = function (version) {
    };
    AbstractMapInoutlet.prototype.memoize = function () {
        var combinator = new MapOutlet.MemoizeMapCombinator();
        combinator.bindInput(this);
        return combinator;
    };
    AbstractMapInoutlet.prototype.filter = function (func) {
        var combinator = new MapOutlet.FilterFieldsCombinator(func);
        combinator.bindInput(this);
        return combinator;
    };
    AbstractMapInoutlet.prototype.map = function (func) {
        if (func.length === 1) {
            var combinator = new Outlet.MapValueCombinator(func);
            combinator.bindInput(this);
            return combinator;
        }
        else {
            var combinator = new MapOutlet.MapFieldValuesCombinator(func);
            combinator.bindInput(this);
            return combinator;
        }
    };
    AbstractMapInoutlet.prototype.reduce = function (identity, accumulator, combiner) {
        var combinator = new MapOutlet.ReduceFieldsCombinator(identity, accumulator, combiner);
        combinator.bindInput(this);
        return combinator;
    };
    AbstractMapInoutlet.prototype.watch = function (func) {
        if (func.length === 1) {
            var combinator = new Outlet.WatchValueCombinator(func);
            combinator.bindInput(this);
            return this;
        }
        else {
            var combinator = new MapOutlet.WatchFieldsCombinator(func);
            combinator.bindInput(this);
            return this;
        }
    };
    return AbstractMapInoutlet;
}());
export { AbstractMapInoutlet };
