// @swim/streamlet/lib/main/AbstractMapInletOutlet.js
import { Cursor } from "@swim/util";
import { BTree } from "@swim/collections";
import { Outlet } from "./Outlet";
import { MapOutlet } from "./MapOutlet";
var AbstractMapInletOutlet = (function () {
    function AbstractMapInletOutlet() {
        this._input = null;
        this._effects = new BTree();
        this._outputs = null;
        this._version = -1;
    }
    AbstractMapInletOutlet.prototype.input = function () {
        return this._input;
    };
    AbstractMapInletOutlet.prototype.bindInput = function (input) {
        if (!MapOutlet.is(input)) {
            throw new TypeError("" + input);
        }
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = input;
        if (this._input !== null) {
            this._input.bindOutput(this);
        }
    };
    AbstractMapInletOutlet.prototype.unbindInput = function () {
        if (this._input !== null) {
            this._input.unbindOutput(this);
        }
        this._input = null;
    };
    AbstractMapInletOutlet.prototype.disconnectInputs = function () {
        if (this._outputs === null) {
            var input = this._input;
            if (input !== null) {
                input.unbindOutput(this);
                this._input = null;
                input.disconnectInputs();
            }
        }
    };
    AbstractMapInletOutlet.prototype.outputIterator = function () {
        return this._outputs !== null ? Cursor.array(this._outputs) : Cursor.empty();
    };
    AbstractMapInletOutlet.prototype.bindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        var newOutputs = new Array(n + 1);
        for (var i = 0; i < n; i += 1) {
            newOutputs[i] = oldOutputs[i];
        }
        newOutputs[n] = output;
        this._outputs = newOutputs;
    };
    AbstractMapInletOutlet.prototype.unbindOutput = function (output) {
        var oldOutputs = this._outputs;
        var n = oldOutputs !== null ? oldOutputs.length : 0;
        for (var i = 0; i < n; i += 1) {
            if (oldOutputs[i] === output) {
                if (n > 1) {
                    var newOutputs = new Array(n - 1);
                    for (var j = 0; j < i; j += 1) {
                        newOutputs[j] = oldOutputs[j];
                    }
                    for (var j = i; j < n - 1; j += 1) {
                        newOutputs[j] = oldOutputs[j + 1];
                    }
                    this._outputs = newOutputs;
                }
                else {
                    this._outputs = null;
                }
                break;
            }
        }
    };
    AbstractMapInletOutlet.prototype.unbindOutputs = function () {
        var outputs = this._outputs;
        if (outputs !== null) {
            this._outputs = null;
            for (var i = 0, n = outputs.length; i < n; i += 1) {
                var output = outputs[i];
                output.unbindInput();
            }
        }
    };
    AbstractMapInletOutlet.prototype.disconnectOutputs = function () {
        if (this._input === null) {
            var outputs = this._outputs;
            if (outputs !== null) {
                this._outputs = null;
                for (var i = 0, n = outputs.length; i < n; i += 1) {
                    var output = outputs[i];
                    output.unbindInput();
                    output.disconnectOutputs();
                }
            }
        }
    };
    AbstractMapInletOutlet.prototype.invalidateOutputKey = function (key, effect) {
        var oldEffects = this._effects;
        if (oldEffects.get(key) !== effect) {
            this.willInvalidateOutputKey(key, effect);
            this._effects = oldEffects.updated(key, effect);
            this._version = -1;
            this.onInvalidateOutputKey(key, effect);
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                var output = this._outputs[i];
                output.invalidateOutput();
            }
            this.didInvalidateOutputKey(key, effect);
        }
    };
    AbstractMapInletOutlet.prototype.invalidateOutput = function () {
        this.invalidate();
    };
    AbstractMapInletOutlet.prototype.invalidateInput = function () {
        this.invalidate();
    };
    AbstractMapInletOutlet.prototype.invalidate = function () {
        if (this._version >= 0) {
            this.willInvalidate();
            this._version = -1;
            this.onInvalidate();
            var n = this._outputs !== null ? this._outputs.length : 0;
            for (var i = 0; i < n; i += 1) {
                this._outputs[i].invalidateOutput();
            }
            this.didInvalidate();
        }
    };
    AbstractMapInletOutlet.prototype.reconcileOutputKey = function (key, version) {
        if (this._version < 0) {
            var oldEffects = this._effects;
            var effect = oldEffects.get(key);
            if (effect !== void 0) {
                this.willReconcileOutputKey(key, effect, version);
                this._effects = oldEffects.removed(key);
                if (this._input !== null) {
                    this._input.reconcileInputKey(key, version);
                }
                this.onReconcileOutputKey(key, effect, version);
                this.didReconcileOutputKey(key, effect, version);
            }
        }
    };
    AbstractMapInletOutlet.prototype.reconcileOutput = function (version) {
        this.reconcile(version);
    };
    AbstractMapInletOutlet.prototype.reconcileInput = function (version) {
        this.reconcile(version);
    };
    AbstractMapInletOutlet.prototype.reconcile = function (version) {
        if (this._version < 0) {
            this.willReconcile(version);
            this._effects.forEach(function (key) {
                this.reconcileOutputKey(key, version);
            }, this);
            this._version = version;
            this.onReconcile(version);
            for (var i = 0, n = this._outputs !== null ? this._outputs.length : 0; i < n; i += 1) {
                this._outputs[i].reconcileOutput(version);
            }
            this.didReconcile(version);
        }
    };
    AbstractMapInletOutlet.prototype.willInvalidateOutputKey = function (key, effect) {
    };
    AbstractMapInletOutlet.prototype.onInvalidateOutputKey = function (key, effect) {
    };
    AbstractMapInletOutlet.prototype.didInvalidateOutputKey = function (key, effect) {
    };
    AbstractMapInletOutlet.prototype.willInvalidate = function () {
    };
    AbstractMapInletOutlet.prototype.onInvalidate = function () {
    };
    AbstractMapInletOutlet.prototype.didInvalidate = function () {
    };
    AbstractMapInletOutlet.prototype.willReconcileOutputKey = function (key, effect, version) {
    };
    AbstractMapInletOutlet.prototype.onReconcileOutputKey = function (key, effect, version) {
    };
    AbstractMapInletOutlet.prototype.didReconcileOutputKey = function (key, effect, version) {
    };
    AbstractMapInletOutlet.prototype.willReconcile = function (version) {
    };
    AbstractMapInletOutlet.prototype.onReconcile = function (version) {
    };
    AbstractMapInletOutlet.prototype.didReconcile = function (version) {
    };
    AbstractMapInletOutlet.prototype.memoize = function () {
        var combinator = new Outlet.MemoizeValueCombinator();
        combinator.bindInput(this);
        return combinator;
    };
    AbstractMapInletOutlet.prototype.map = function (func) {
        var combinator = new Outlet.MapValueCombinator(func);
        combinator.bindInput(this);
        return combinator;
    };
    AbstractMapInletOutlet.prototype.watch = function (func) {
        var combinator = new Outlet.WatchValueCombinator(func);
        combinator.bindInput(this);
        return this;
    };
    return AbstractMapInletOutlet;
}());
export { AbstractMapInletOutlet };
