// @swim/streamlet/lib/main/StreamletInoutlet.js
import { __extends } from "tslib";
import { AbstractInoutlet } from "./AbstractInoutlet";
var StreamletInoutlet = (function (_super) {
    __extends(StreamletInoutlet, _super);
    function StreamletInoutlet(streamlet) {
        var _this = _super.call(this) || this;
        _this._streamlet = streamlet;
        return _this;
    }
    StreamletInoutlet.prototype.streamlet = function () {
        return this._streamlet;
    };
    StreamletInoutlet.prototype.get = function () {
        var streamlet = this._streamlet;
        if (streamlet.getOutput) {
            var output = streamlet.getOutput(this);
            if (output !== void 0) {
                return output;
            }
        }
        if (this._input !== null) {
            return this._input.get();
        }
        return void 0;
    };
    StreamletInoutlet.prototype.willInvalidate = function () {
        var streamlet = this._streamlet;
        if (streamlet.willInvalidateOutlet) {
            streamlet.willInvalidateOutlet(this);
        }
    };
    StreamletInoutlet.prototype.didInvalidate = function () {
        var streamlet = this._streamlet;
        if (streamlet.didInvalidateOutlet) {
            streamlet.didInvalidateOutlet(this);
        }
        else {
            streamlet.invalidate();
        }
    };
    StreamletInoutlet.prototype.willUpdate = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.willReconcileOutlet) {
            streamlet.willReconcileOutlet(this, version);
        }
    };
    StreamletInoutlet.prototype.didUpdate = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.didReconcileOutlet) {
            streamlet.didReconcileOutlet(this, version);
        }
    };
    return StreamletInoutlet;
}(AbstractInoutlet));
export { StreamletInoutlet };
