// @swim/streamlet/lib/main/StreamletInlet.js
import { __extends } from "tslib";
import { AbstractInlet } from "./AbstractInlet";
var StreamletInlet = (function (_super) {
    __extends(StreamletInlet, _super);
    function StreamletInlet(streamlet) {
        var _this = _super.call(this) || this;
        _this._streamlet = streamlet;
        return _this;
    }
    StreamletInlet.prototype.streamlet = function () {
        return this._streamlet;
    };
    StreamletInlet.prototype.willInvalidateOutput = function () {
        var streamlet = this._streamlet;
        if (streamlet.willInvalidateInlet) {
            streamlet.willInvalidateInlet(this);
        }
    };
    StreamletInlet.prototype.didInvalidateOutput = function () {
        var streamlet = this._streamlet;
        if (streamlet.didInvalidateInlet) {
            streamlet.didInvalidateInlet(this);
        }
        else {
            streamlet.invalidate();
        }
    };
    StreamletInlet.prototype.willReconcileOutput = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.willReconcileInlet) {
            streamlet.willReconcileInlet(this, version);
        }
    };
    StreamletInlet.prototype.didReconcileOutput = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.didReconcileInlet) {
            streamlet.didReconcileInlet(this, version);
        }
        else {
            streamlet.reconcile(version);
        }
    };
    return StreamletInlet;
}(AbstractInlet));
export { StreamletInlet };
