// @swim/streamlet/lib/main/StreamletOutlet.js
import { __extends } from "tslib";
import { AbstractOutlet } from "./AbstractOutlet";
var StreamletOutlet = (function (_super) {
    __extends(StreamletOutlet, _super);
    function StreamletOutlet(streamlet) {
        var _this = _super.call(this) || this;
        _this._streamlet = streamlet;
        return _this;
    }
    StreamletOutlet.prototype.streamlet = function () {
        return this._streamlet;
    };
    StreamletOutlet.prototype.get = function () {
        var streamlet = this._streamlet;
        if (streamlet.getOutput) {
            return streamlet.getOutput(this);
        }
        return void 0;
    };
    StreamletOutlet.prototype.willInvalidateInput = function () {
        var streamlet = this._streamlet;
        if (streamlet.willInvalidateOutlet) {
            streamlet.willInvalidateOutlet(this);
        }
    };
    StreamletOutlet.prototype.didInvalidateInput = function () {
        var streamlet = this._streamlet;
        if (streamlet.didInvalidateOutlet) {
            streamlet.didInvalidateOutlet(this);
        }
        else {
            streamlet.invalidate();
        }
    };
    StreamletOutlet.prototype.willReconcileInput = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.willReconcileOutlet) {
            streamlet.willReconcileOutlet(this, version);
        }
    };
    StreamletOutlet.prototype.didReconcileInput = function (version) {
        var streamlet = this._streamlet;
        if (streamlet.didReconcileOutlet) {
            streamlet.didReconcileOutlet(this, version);
        }
    };
    return StreamletOutlet;
}(AbstractOutlet));
export { StreamletOutlet };
