// @swim/structure/lib/main/selector/ChildrenSelector.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { Selector } from "../Selector";
import { Interpreter } from "../Interpreter";
var ChildrenSelector = (function (_super) {
    __extends(ChildrenSelector, _super);
    function ChildrenSelector(then) {
        var _this = _super.call(this) || this;
        _this._then = then;
        return _this;
    }
    ChildrenSelector.prototype.then = function () {
        return this._then;
    };
    ChildrenSelector.prototype.forSelected = function (interpreter, callback, thisArg) {
        var selected;
        interpreter.willSelect(this);
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope().toValue();
            if (scope instanceof Item.Record) {
                var children = scope.iterator();
                while (selected === void 0 && children.hasNext()) {
                    var child = children.next().value;
                    interpreter.pushScope(child);
                    selected = this._then.forSelected(interpreter, callback, thisArg);
                    interpreter.popScope();
                }
            }
            interpreter.pushScope(scope);
        }
        interpreter.didSelect(this, selected);
        return selected;
    };
    ChildrenSelector.prototype.mapSelected = function (interpreter, transform, thisArg) {
        var result;
        interpreter.willTransform(this);
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope().toValue();
            if (scope instanceof Item.Record) {
                var children = scope.iterator();
                while (children.hasNext()) {
                    var oldChild = children.next().value;
                    interpreter.pushScope(oldChild);
                    var newChild = this._then.mapSelected(interpreter, transform);
                    interpreter.popScope();
                    if (newChild.isDefined()) {
                        if (newChild !== oldChild) {
                            children.set(newChild);
                        }
                    }
                    else {
                        children.delete();
                    }
                }
            }
            interpreter.pushScope(scope);
            result = scope;
        }
        else {
            result = Item.absent();
        }
        interpreter.didTransform(this, result);
        return result;
    };
    ChildrenSelector.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var then = this._then.substitute(interpreter);
        if (!(then instanceof Selector)) {
            then = this._then;
        }
        return new ChildrenSelector(then);
    };
    ChildrenSelector.prototype.andThen = function (then) {
        return new ChildrenSelector(this._then.andThen(then));
    };
    ChildrenSelector.prototype.typeOrder = function () {
        return 17;
    };
    ChildrenSelector.prototype.compareTo = function (that) {
        if (that instanceof ChildrenSelector) {
            return this._then.compareTo(that._then);
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    ChildrenSelector.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof ChildrenSelector) {
            return this._then.equals(that._then);
        }
        return false;
    };
    ChildrenSelector.prototype.hashCode = function () {
        if (ChildrenSelector._hashSeed === void 0) {
            ChildrenSelector._hashSeed = Murmur3.seed(ChildrenSelector);
        }
        return Murmur3.mash(Murmur3.mix(ChildrenSelector._hashSeed, this._then.hashCode()));
    };
    ChildrenSelector.prototype.debugThen = function (output) {
        output = output.write(46).write("children").write(40).write(41);
        this._then.debugThen(output);
    };
    ChildrenSelector.prototype.clone = function () {
        return new ChildrenSelector(this._then.clone());
    };
    return ChildrenSelector;
}(Selector));
export { ChildrenSelector };
Item.ChildrenSelector = ChildrenSelector;
