// @swim/structure/lib/main/selector/GetSelector.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { Selector } from "../Selector";
import { Interpreter } from "../Interpreter";
var GetSelector = (function (_super) {
    __extends(GetSelector, _super);
    function GetSelector(key, then) {
        var _this = _super.call(this) || this;
        _this._key = key.commit();
        _this._then = then;
        return _this;
    }
    GetSelector.prototype.accessor = function () {
        return this._key;
    };
    GetSelector.prototype.then = function () {
        return this._then;
    };
    GetSelector.prototype.forSelected = function (interpreter, callback, thisArg) {
        interpreter.willSelect(this);
        var key = this._key.evaluate(interpreter).toValue();
        var selected = GetSelector.forSelected(key, this._then, interpreter, callback, thisArg);
        interpreter.didSelect(this, selected);
        return selected;
    };
    GetSelector.forSelected = function (key, then, interpreter, callback, thisArg) {
        var selected;
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope().toValue();
            var field = void 0;
            if (scope instanceof Item.Record) {
                field = scope.getField(key);
                if (field) {
                    interpreter.pushScope(field.toValue());
                    selected = then.forSelected(interpreter, callback, thisArg);
                    interpreter.popScope();
                }
            }
            if (field === void 0 && selected === void 0) {
                GetSelector.forSelected(key, then, interpreter, callback, thisArg);
            }
            interpreter.pushScope(scope);
        }
        return selected;
    };
    GetSelector.prototype.mapSelected = function (interpreter, transform, thisArg) {
        var result;
        interpreter.willTransform(this);
        var key = this._key.evaluate(interpreter).toValue();
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope().toValue();
            if (scope instanceof Item.Record) {
                var oldField = scope.getField(key);
                if (oldField) {
                    interpreter.pushScope(oldField.toValue());
                    var newItem = this._then.mapSelected(interpreter, transform, thisArg);
                    interpreter.popScope();
                    if (newItem instanceof Item.Field) {
                        if (key.equals(newItem.key)) {
                            scope.set(key, newItem.toValue());
                        }
                        else {
                            scope.delete(key);
                            scope.push(newItem);
                        }
                    }
                    else if (newItem.isDefined()) {
                        scope.set(key, newItem.toValue());
                    }
                    else {
                        scope.delete(key);
                    }
                }
            }
            interpreter.pushScope(scope);
            result = scope;
        }
        else {
            result = Item.absent();
        }
        interpreter.didTransform(this, result);
        return result;
    };
    GetSelector.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var key = this._key.evaluate(interpreter).toValue();
        var value = GetSelector.substitute(key, this._then, interpreter);
        if (value) {
            return value;
        }
        var then = this._then.substitute(interpreter);
        if (!(then instanceof Selector)) {
            then = this._then;
        }
        return new GetSelector(this._key, then);
    };
    GetSelector.substitute = function (key, then, interpreter) {
        var selected;
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope().toValue();
            var field = void 0;
            if (scope instanceof Item.Record) {
                field = scope.getField(key);
                if (field) {
                    selected = field.toValue().substitute(interpreter);
                }
            }
            if (field === void 0 && selected === void 0) {
                GetSelector.substitute(key, then, interpreter);
            }
            interpreter.pushScope(scope);
        }
        return selected;
    };
    GetSelector.prototype.andThen = function (then) {
        return new GetSelector(this._key, this._then.andThen(then));
    };
    GetSelector.prototype.typeOrder = function () {
        return 12;
    };
    GetSelector.prototype.compareTo = function (that) {
        if (that instanceof GetSelector) {
            var order = this._key.compareTo(that._key);
            if (order === 0) {
                order = this._then.compareTo(that._then);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    GetSelector.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof GetSelector) {
            return this._key.equals(that._key) && this._then.equals(that._then);
        }
        return false;
    };
    GetSelector.prototype.hashCode = function () {
        if (GetSelector._hashSeed === void 0) {
            GetSelector._hashSeed = Murmur3.seed(GetSelector);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(GetSelector._hashSeed, this._key.hashCode()), this._then.hashCode()));
    };
    GetSelector.prototype.debugThen = function (output) {
        output = output.write(46).write("get").write(40).debug(this._key).write(41);
        this._then.debugThen(output);
    };
    GetSelector.prototype.clone = function () {
        return new GetSelector(this._key.clone(), this._then.clone());
    };
    return GetSelector;
}(Selector));
export { GetSelector };
Item.GetSelector = GetSelector;
