// @swim/structure/lib/main/selector/KeysSelector.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { Selector } from "../Selector";
import { Interpreter } from "../Interpreter";
var KeysSelector = (function (_super) {
    __extends(KeysSelector, _super);
    function KeysSelector(then) {
        var _this = _super.call(this) || this;
        _this._then = then;
        return _this;
    }
    KeysSelector.prototype.then = function () {
        return this._then;
    };
    KeysSelector.prototype.forSelected = function (interpreter, callback, thisArg) {
        var selected;
        interpreter.willSelect(this);
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope();
            if (scope instanceof Item.Record) {
                var children = scope.iterator();
                while (selected === void 0 && children.hasNext()) {
                    var child = children.next().value;
                    if (child instanceof Item.Field) {
                        interpreter.pushScope(child.key);
                        selected = this._then.forSelected(interpreter, callback, thisArg);
                        interpreter.popScope();
                    }
                }
            }
            else if (scope instanceof Item.Field) {
                interpreter.pushScope(scope.key);
                selected = this._then.forSelected(interpreter, callback, thisArg);
                interpreter.popScope();
            }
            interpreter.pushScope(scope);
        }
        interpreter.didSelect(this, selected);
        return selected;
    };
    KeysSelector.prototype.mapSelected = function (interpreter, transform, thisArg) {
        var result;
        interpreter.willTransform(this);
        if (interpreter.scopeDepth() !== 0) {
            var scope = interpreter.popScope();
            if (scope instanceof Item.Record) {
                var children = scope.iterator();
                while (children.hasNext()) {
                    var child = children.next().value;
                    if (child instanceof Item.Field) {
                        var oldKey = child.key;
                        interpreter.pushScope(oldKey);
                        var newKey = this._then.mapSelected(interpreter, transform, thisArg).toValue();
                        interpreter.popScope();
                        if (newKey.isDefined()) {
                            if (oldKey !== newKey) {
                                if (scope instanceof Item.Attr && newKey instanceof Item.Text) {
                                    children.set(Item.Attr.of(newKey, scope.toValue()));
                                }
                                else {
                                    children.set(Item.Slot.of(newKey, scope.toValue()));
                                }
                            }
                        }
                        else {
                            children.delete();
                        }
                    }
                }
            }
            else if (scope instanceof Item.Field) {
                var oldKey = scope.key;
                interpreter.pushScope(oldKey);
                var newKey = this._then.mapSelected(interpreter, transform, thisArg).toValue();
                interpreter.popScope();
                if (newKey.isDefined()) {
                    if (oldKey !== newKey) {
                        if (scope instanceof Item.Attr && newKey instanceof Item.Text) {
                            scope = Item.Attr.of(newKey, scope.toValue());
                        }
                        else {
                            scope = Item.Slot.of(newKey, scope.toValue());
                        }
                    }
                }
                else {
                    scope = Item.absent();
                }
            }
            interpreter.pushScope(scope);
            result = scope;
        }
        else {
            result = Item.absent();
        }
        interpreter.didTransform(this, result);
        return result;
    };
    KeysSelector.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var then = this._then.substitute(interpreter);
        if (!(then instanceof Selector)) {
            then = this._then;
        }
        return new KeysSelector(then);
    };
    KeysSelector.prototype.andThen = function (then) {
        return new KeysSelector(this._then.andThen(then));
    };
    KeysSelector.prototype.typeOrder = function () {
        return 15;
    };
    KeysSelector.prototype.compareTo = function (that) {
        if (that instanceof KeysSelector) {
            return this._then.compareTo(that._then);
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    KeysSelector.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof KeysSelector) {
            return this._then.equals(that._then);
        }
        return false;
    };
    KeysSelector.prototype.hashCode = function () {
        if (KeysSelector._hashSeed === void 0) {
            KeysSelector._hashSeed = Murmur3.seed(KeysSelector);
        }
        return Murmur3.mash(Murmur3.mix(KeysSelector._hashSeed, this._then.hashCode()));
    };
    KeysSelector.prototype.debugThen = function (output) {
        output = output.write(46).write("keys").write(40).write(41);
        this._then.debugThen(output);
    };
    KeysSelector.prototype.clone = function () {
        return new KeysSelector(this._then.clone());
    };
    return KeysSelector;
}(Selector));
export { KeysSelector };
Item.KeysSelector = KeysSelector;
