// @swim/structure/lib/main/operator/AndOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var AndOperator = (function (_super) {
    __extends(AndOperator, _super);
    function AndOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    AndOperator.prototype.operator = function () {
        return "&&";
    };
    AndOperator.prototype.precedence = function () {
        return 4;
    };
    AndOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var result;
        var argument1 = this._operand1.evaluate(interpreter);
        if (argument1.booleanValue(false)) {
            var argument2 = this._operand2.evaluate(interpreter);
            result = argument2;
        }
        else {
            result = argument1;
        }
        interpreter.didOperate(this, result);
        return result;
    };
    AndOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.and(argument2);
    };
    AndOperator.prototype.typeOrder = function () {
        return 22;
    };
    AndOperator.prototype.compareTo = function (that) {
        if (that instanceof AndOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    AndOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof AndOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    AndOperator.prototype.hashCode = function () {
        if (AndOperator._hashSeed === void 0) {
            AndOperator._hashSeed = Murmur3.seed(AndOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(AndOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    AndOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("and").write(40)
            .debug(this._operand2).write(41);
    };
    AndOperator.prototype.clone = function () {
        return new AndOperator(this._operand1.clone(), this._operand2.clone());
    };
    return AndOperator;
}(BinaryOperator));
export { AndOperator };
Item.AndOperator = AndOperator;
