// @swim/structure/lib/main/operator/BitwiseOrOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var BitwiseOrOperator = (function (_super) {
    __extends(BitwiseOrOperator, _super);
    function BitwiseOrOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    BitwiseOrOperator.prototype.operator = function () {
        return "|";
    };
    BitwiseOrOperator.prototype.precedence = function () {
        return 5;
    };
    BitwiseOrOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.bitwiseOr(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    BitwiseOrOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.bitwiseOr(argument2);
    };
    BitwiseOrOperator.prototype.typeOrder = function () {
        return 23;
    };
    BitwiseOrOperator.prototype.compareTo = function (that) {
        if (that instanceof BitwiseOrOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    BitwiseOrOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof BitwiseOrOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    BitwiseOrOperator.prototype.hashCode = function () {
        if (BitwiseOrOperator._hashSeed === void 0) {
            BitwiseOrOperator._hashSeed = Murmur3.seed(BitwiseOrOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(BitwiseOrOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    BitwiseOrOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("bitwiseOr").write(40)
            .debug(this._operand2).write(41);
    };
    BitwiseOrOperator.prototype.clone = function () {
        return new BitwiseOrOperator(this._operand1.clone(), this._operand2.clone());
    };
    return BitwiseOrOperator;
}(BinaryOperator));
export { BitwiseOrOperator };
Item.BitwiseOrOperator = BitwiseOrOperator;
