// @swim/structure/lib/main/operator/BitwiseNotOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { UnaryOperator } from "./UnaryOperator";
import { Interpreter } from "../Interpreter";
var BitwiseNotOperator = (function (_super) {
    __extends(BitwiseNotOperator, _super);
    function BitwiseNotOperator(operand) {
        return _super.call(this, operand) || this;
    }
    BitwiseNotOperator.prototype.operator = function () {
        return "~";
    };
    BitwiseNotOperator.prototype.precedence = function () {
        return 10;
    };
    BitwiseNotOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.evaluate(interpreter);
        return argument.bitwiseNot();
    };
    BitwiseNotOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.substitute(interpreter);
        return argument.bitwiseNot();
    };
    BitwiseNotOperator.prototype.typeOrder = function () {
        return 38;
    };
    BitwiseNotOperator.prototype.compareTo = function (that) {
        if (that instanceof BitwiseNotOperator) {
            return this._operand.compareTo(that._operand);
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    BitwiseNotOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof BitwiseNotOperator) {
            return this._operand.equals(that._operand);
        }
        return false;
    };
    BitwiseNotOperator.prototype.hashCode = function () {
        if (BitwiseNotOperator._hashSeed === void 0) {
            BitwiseNotOperator._hashSeed = Murmur3.seed(BitwiseNotOperator);
        }
        return Murmur3.mash(Murmur3.mix(BitwiseNotOperator._hashSeed, this._operand.hashCode()));
    };
    BitwiseNotOperator.prototype.debug = function (output) {
        output.debug(this._operand).write(46).write("bitwiseNot").write(40).write(41);
    };
    BitwiseNotOperator.prototype.clone = function () {
        return new BitwiseNotOperator(this._operand.clone());
    };
    return BitwiseNotOperator;
}(UnaryOperator));
export { BitwiseNotOperator };
Item.BitwiseNotOperator = BitwiseNotOperator;
