// @swim/structure/lib/main/operator/BitwiseAndOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var BitwiseAndOperator = (function (_super) {
    __extends(BitwiseAndOperator, _super);
    function BitwiseAndOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    BitwiseAndOperator.prototype.operator = function () {
        return "&";
    };
    BitwiseAndOperator.prototype.precedence = function () {
        return 7;
    };
    BitwiseAndOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.bitwiseAnd(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    BitwiseAndOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.bitwiseAnd(argument2);
    };
    BitwiseAndOperator.prototype.typeOrder = function () {
        return 25;
    };
    BitwiseAndOperator.prototype.compareTo = function (that) {
        if (that instanceof BitwiseAndOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    BitwiseAndOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof BitwiseAndOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    BitwiseAndOperator.prototype.hashCode = function () {
        if (BitwiseAndOperator._hashSeed === void 0) {
            BitwiseAndOperator._hashSeed = Murmur3.seed(BitwiseAndOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(BitwiseAndOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    BitwiseAndOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("bitwiseAnd").write(40)
            .debug(this._operand2).write(41);
    };
    BitwiseAndOperator.prototype.clone = function () {
        return new BitwiseAndOperator(this._operand1.clone(), this._operand2.clone());
    };
    return BitwiseAndOperator;
}(BinaryOperator));
export { BitwiseAndOperator };
Item.BitwiseAndOperator = BitwiseAndOperator;
