// @swim/structure/lib/main/operator/DivideOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var DivideOperator = (function (_super) {
    __extends(DivideOperator, _super);
    function DivideOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    DivideOperator.prototype.operator = function () {
        return "/";
    };
    DivideOperator.prototype.precedence = function () {
        return 9;
    };
    DivideOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.divide(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    DivideOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.divide(argument2);
    };
    DivideOperator.prototype.typeOrder = function () {
        return 35;
    };
    DivideOperator.prototype.compareTo = function (that) {
        if (that instanceof DivideOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    DivideOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof DivideOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    DivideOperator.prototype.hashCode = function () {
        if (DivideOperator._hashSeed === void 0) {
            DivideOperator._hashSeed = Murmur3.seed(DivideOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(DivideOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    DivideOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("divide").write(40)
            .debug(this._operand2).write(41);
    };
    DivideOperator.prototype.clone = function () {
        return new DivideOperator(this._operand1.clone(), this._operand2.clone());
    };
    return DivideOperator;
}(BinaryOperator));
export { DivideOperator };
Item.DivideOperator = DivideOperator;
