// @swim/structure/lib/main/operator/EqOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var EqOperator = (function (_super) {
    __extends(EqOperator, _super);
    function EqOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    EqOperator.prototype.operator = function () {
        return "==";
    };
    EqOperator.prototype.precedence = function () {
        return 0;
    };
    EqOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.eq(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    EqOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.eq(argument2);
    };
    EqOperator.prototype.typeOrder = function () {
        return 28;
    };
    EqOperator.prototype.compareTo = function (that) {
        if (that instanceof EqOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    EqOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof EqOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    EqOperator.prototype.hashCode = function () {
        if (EqOperator._hashSeed === void 0) {
            EqOperator._hashSeed = Murmur3.seed(EqOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(EqOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    EqOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("eq").write(40)
            .debug(this._operand2).write(41);
    };
    EqOperator.prototype.clone = function () {
        return new EqOperator(this._operand1.clone(), this._operand2.clone());
    };
    return EqOperator;
}(BinaryOperator));
export { EqOperator };
Item.EqOperator = EqOperator;
