// @swim/structure/lib/main/operator/GeOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var GeOperator = (function (_super) {
    __extends(GeOperator, _super);
    function GeOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    GeOperator.prototype.operator = function () {
        return ">=";
    };
    GeOperator.prototype.precedence = function () {
        return 0;
    };
    GeOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.ge(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    GeOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.ge(argument2);
    };
    GeOperator.prototype.typeOrder = function () {
        return 30;
    };
    GeOperator.prototype.compareTo = function (that) {
        if (that instanceof GeOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    GeOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof GeOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    GeOperator.prototype.hashCode = function () {
        if (GeOperator._hashSeed === void 0) {
            GeOperator._hashSeed = Murmur3.seed(GeOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(GeOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    GeOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("ge").write(40)
            .debug(this._operand2).write(41);
    };
    GeOperator.prototype.clone = function () {
        return new GeOperator(this._operand1.clone(), this._operand2.clone());
    };
    return GeOperator;
}(BinaryOperator));
export { GeOperator };
Item.GeOperator = GeOperator;
