// @swim/structure/lib/main/operator/GtOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var GtOperator = (function (_super) {
    __extends(GtOperator, _super);
    function GtOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    GtOperator.prototype.operator = function () {
        return ">";
    };
    GtOperator.prototype.precedence = function () {
        return 0;
    };
    GtOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.gt(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    GtOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.gt(argument2);
    };
    GtOperator.prototype.typeOrder = function () {
        return 31;
    };
    GtOperator.prototype.compareTo = function (that) {
        if (that instanceof GtOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    GtOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof GtOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    GtOperator.prototype.hashCode = function () {
        if (GtOperator._hashSeed === void 0) {
            GtOperator._hashSeed = Murmur3.seed(GtOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(GtOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    GtOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("gt").write(40)
            .debug(this._operand2).write(41);
    };
    GtOperator.prototype.clone = function () {
        return new GtOperator(this._operand1.clone(), this._operand2.clone());
    };
    return GtOperator;
}(BinaryOperator));
export { GtOperator };
Item.GtOperator = GtOperator;
