// @swim/structure/lib/main/operator/LeOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var LeOperator = (function (_super) {
    __extends(LeOperator, _super);
    function LeOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    LeOperator.prototype.operator = function () {
        return "<=";
    };
    LeOperator.prototype.precedence = function () {
        return 0;
    };
    LeOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.le(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    LeOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.le(argument2);
    };
    LeOperator.prototype.typeOrder = function () {
        return 27;
    };
    LeOperator.prototype.compareTo = function (that) {
        if (that instanceof LeOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    LeOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof LeOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    LeOperator.prototype.hashCode = function () {
        if (LeOperator._hashSeed === void 0) {
            LeOperator._hashSeed = Murmur3.seed(LeOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(LeOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    LeOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("le").write(40)
            .debug(this._operand2).write(41);
    };
    LeOperator.prototype.clone = function () {
        return new LeOperator(this._operand1.clone(), this._operand2.clone());
    };
    return LeOperator;
}(BinaryOperator));
export { LeOperator };
Item.LeOperator = LeOperator;
