// @swim/structure/lib/main/operator/LtOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var LtOperator = (function (_super) {
    __extends(LtOperator, _super);
    function LtOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    LtOperator.prototype.operator = function () {
        return "<";
    };
    LtOperator.prototype.precedence = function () {
        return 0;
    };
    LtOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.lt(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    LtOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.lt(argument2);
    };
    LtOperator.prototype.typeOrder = function () {
        return 26;
    };
    LtOperator.prototype.compareTo = function (that) {
        if (that instanceof LtOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    LtOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof LtOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    LtOperator.prototype.hashCode = function () {
        if (LtOperator._hashSeed === void 0) {
            LtOperator._hashSeed = Murmur3.seed(LtOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(LtOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    LtOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("lt").write(40)
            .debug(this._operand2).write(41);
    };
    LtOperator.prototype.clone = function () {
        return new LtOperator(this._operand1.clone(), this._operand2.clone());
    };
    return LtOperator;
}(BinaryOperator));
export { LtOperator };
Item.LtOperator = LtOperator;
