// @swim/structure/lib/main/operator/MinusOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var MinusOperator = (function (_super) {
    __extends(MinusOperator, _super);
    function MinusOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    MinusOperator.prototype.operator = function () {
        return "-";
    };
    MinusOperator.prototype.precedence = function () {
        return 8;
    };
    MinusOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.minus(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    MinusOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.minus(argument2);
    };
    MinusOperator.prototype.typeOrder = function () {
        return 33;
    };
    MinusOperator.prototype.compareTo = function (that) {
        if (that instanceof MinusOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    MinusOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof MinusOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    MinusOperator.prototype.hashCode = function () {
        if (MinusOperator._hashSeed === void 0) {
            MinusOperator._hashSeed = Murmur3.seed(MinusOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(MinusOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    MinusOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("minus").write(40)
            .debug(this._operand2).write(41);
    };
    MinusOperator.prototype.clone = function () {
        return new MinusOperator(this._operand1.clone(), this._operand2.clone());
    };
    return MinusOperator;
}(BinaryOperator));
export { MinusOperator };
Item.MinusOperator = MinusOperator;
