// @swim/structure/lib/main/operator/ModuloOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var ModuloOperator = (function (_super) {
    __extends(ModuloOperator, _super);
    function ModuloOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    ModuloOperator.prototype.operator = function () {
        return "%";
    };
    ModuloOperator.prototype.precedence = function () {
        return 9;
    };
    ModuloOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.modulo(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    ModuloOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.modulo(argument2);
    };
    ModuloOperator.prototype.typeOrder = function () {
        return 36;
    };
    ModuloOperator.prototype.compareTo = function (that) {
        if (that instanceof ModuloOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    ModuloOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof ModuloOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    ModuloOperator.prototype.hashCode = function () {
        if (ModuloOperator._hashSeed === void 0) {
            ModuloOperator._hashSeed = Murmur3.seed(ModuloOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(ModuloOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    ModuloOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("modulo").write(40)
            .debug(this._operand2).write(41);
    };
    ModuloOperator.prototype.clone = function () {
        return new ModuloOperator(this._operand1.clone(), this._operand2.clone());
    };
    return ModuloOperator;
}(BinaryOperator));
export { ModuloOperator };
Item.ModuloOperator = ModuloOperator;
