// @swim/structure/lib/main/operator/NegativeOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { UnaryOperator } from "./UnaryOperator";
import { Interpreter } from "../Interpreter";
var NegativeOperator = (function (_super) {
    __extends(NegativeOperator, _super);
    function NegativeOperator(operand) {
        return _super.call(this, operand) || this;
    }
    NegativeOperator.prototype.operator = function () {
        return "-";
    };
    NegativeOperator.prototype.precedence = function () {
        return 10;
    };
    NegativeOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.evaluate(interpreter);
        return argument.negative();
    };
    NegativeOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.substitute(interpreter);
        return argument.negative();
    };
    NegativeOperator.prototype.typeOrder = function () {
        return 39;
    };
    NegativeOperator.prototype.compareTo = function (that) {
        if (that instanceof NegativeOperator) {
            return this._operand.compareTo(that._operand);
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    NegativeOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof NegativeOperator) {
            return this._operand.equals(that._operand);
        }
        return false;
    };
    NegativeOperator.prototype.hashCode = function () {
        if (NegativeOperator._hashSeed === void 0) {
            NegativeOperator._hashSeed = Murmur3.seed(NegativeOperator);
        }
        return Murmur3.mash(Murmur3.mix(NegativeOperator._hashSeed, this._operand.hashCode()));
    };
    NegativeOperator.prototype.debug = function (output) {
        output.debug(this._operand).write(46).write("negative").write(40).write(41);
    };
    NegativeOperator.prototype.clone = function () {
        return new NegativeOperator(this._operand.clone());
    };
    return NegativeOperator;
}(UnaryOperator));
export { NegativeOperator };
Item.NegativeOperator = NegativeOperator;
