// @swim/structure/lib/main/operator/NeOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var NeOperator = (function (_super) {
    __extends(NeOperator, _super);
    function NeOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    NeOperator.prototype.operator = function () {
        return "!=";
    };
    NeOperator.prototype.precedence = function () {
        return 0;
    };
    NeOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.ne(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    NeOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.ne(argument2);
    };
    NeOperator.prototype.typeOrder = function () {
        return 29;
    };
    NeOperator.prototype.compareTo = function (that) {
        if (that instanceof NeOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    NeOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof NeOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    NeOperator.prototype.hashCode = function () {
        if (NeOperator._hashSeed === void 0) {
            NeOperator._hashSeed = Murmur3.seed(NeOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(NeOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    NeOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("ne").write(40)
            .debug(this._operand2).write(41);
    };
    NeOperator.prototype.clone = function () {
        return new NeOperator(this._operand1.clone(), this._operand2.clone());
    };
    return NeOperator;
}(BinaryOperator));
export { NeOperator };
Item.NeOperator = NeOperator;
