// @swim/structure/lib/main/operator/NotOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { UnaryOperator } from "./UnaryOperator";
import { Interpreter } from "../Interpreter";
var NotOperator = (function (_super) {
    __extends(NotOperator, _super);
    function NotOperator(operand) {
        return _super.call(this, operand) || this;
    }
    NotOperator.prototype.operator = function () {
        return "!";
    };
    NotOperator.prototype.precedence = function () {
        return 10;
    };
    NotOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.evaluate(interpreter);
        return argument.not();
    };
    NotOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument = this._operand.substitute(interpreter);
        return argument.not();
    };
    NotOperator.prototype.typeOrder = function () {
        return 37;
    };
    NotOperator.prototype.compareTo = function (that) {
        if (that instanceof NotOperator) {
            return this._operand.compareTo(that._operand);
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    NotOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof NotOperator) {
            return this._operand.equals(that._operand);
        }
        return false;
    };
    NotOperator.prototype.hashCode = function () {
        if (NotOperator._hashSeed === void 0) {
            NotOperator._hashSeed = Murmur3.seed(NotOperator);
        }
        return Murmur3.mash(Murmur3.mix(NotOperator._hashSeed, this._operand.hashCode()));
    };
    NotOperator.prototype.debug = function (output) {
        output.debug(this._operand).write(46).write("not").write(40).write(41);
    };
    NotOperator.prototype.clone = function () {
        return new NotOperator(this._operand.clone());
    };
    return NotOperator;
}(UnaryOperator));
export { NotOperator };
Item.NotOperator = NotOperator;
