// @swim/structure/lib/main/operator/OrOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var OrOperator = (function (_super) {
    __extends(OrOperator, _super);
    function OrOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    OrOperator.prototype.operator = function () {
        return "||";
    };
    OrOperator.prototype.precedence = function () {
        return 3;
    };
    OrOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var result;
        var argument1 = this._operand1.evaluate(interpreter);
        if (argument1.booleanValue(false)) {
            result = argument1;
        }
        else {
            var argument2 = this._operand2.evaluate(interpreter);
            result = argument2;
        }
        interpreter.didOperate(this, result);
        return result;
    };
    OrOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.or(argument2);
    };
    OrOperator.prototype.typeOrder = function () {
        return 21;
    };
    OrOperator.prototype.compareTo = function (that) {
        if (that instanceof OrOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    OrOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof OrOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    OrOperator.prototype.hashCode = function () {
        if (OrOperator._hashSeed === void 0) {
            OrOperator._hashSeed = Murmur3.seed(OrOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(OrOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    OrOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("or").write(40)
            .debug(this._operand2).write(41);
    };
    OrOperator.prototype.clone = function () {
        return new OrOperator(this._operand1.clone(), this._operand2.clone());
    };
    return OrOperator;
}(BinaryOperator));
export { OrOperator };
Item.OrOperator = OrOperator;
