// @swim/structure/lib/main/operator/PlusOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var PlusOperator = (function (_super) {
    __extends(PlusOperator, _super);
    function PlusOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    PlusOperator.prototype.operator = function () {
        return "+";
    };
    PlusOperator.prototype.precedence = function () {
        return 8;
    };
    PlusOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.plus(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    PlusOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.plus(argument2);
    };
    PlusOperator.prototype.typeOrder = function () {
        return 32;
    };
    PlusOperator.prototype.compareTo = function (that) {
        if (that instanceof PlusOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    PlusOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof PlusOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    PlusOperator.prototype.hashCode = function () {
        if (PlusOperator._hashSeed === void 0) {
            PlusOperator._hashSeed = Murmur3.seed(PlusOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(PlusOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    PlusOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("plus").write(40)
            .debug(this._operand2).write(41);
    };
    PlusOperator.prototype.clone = function () {
        return new PlusOperator(this._operand1.clone(), this._operand2.clone());
    };
    return PlusOperator;
}(BinaryOperator));
export { PlusOperator };
Item.PlusOperator = PlusOperator;
