// @swim/structure/lib/main/operator/TimesOperator.js
import { __extends } from "tslib";
import { Murmur3, Objects } from "@swim/util";
import { Item } from "../Item";
import { BinaryOperator } from "./BinaryOperator";
import { Interpreter } from "../Interpreter";
var TimesOperator = (function (_super) {
    __extends(TimesOperator, _super);
    function TimesOperator(operand1, operand2) {
        return _super.call(this, operand1, operand2) || this;
    }
    TimesOperator.prototype.operator = function () {
        return "*";
    };
    TimesOperator.prototype.precedence = function () {
        return 9;
    };
    TimesOperator.prototype.evaluate = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        interpreter.willOperate(this);
        var argument1 = this._operand1.evaluate(interpreter);
        var argument2 = this._operand2.evaluate(interpreter);
        var result = argument1.times(argument2);
        interpreter.didOperate(this, result);
        return result;
    };
    TimesOperator.prototype.substitute = function (interpreter) {
        interpreter = Interpreter.fromAny(interpreter);
        var argument1 = this._operand1.substitute(interpreter);
        var argument2 = this._operand2.substitute(interpreter);
        return argument1.times(argument2);
    };
    TimesOperator.prototype.typeOrder = function () {
        return 34;
    };
    TimesOperator.prototype.compareTo = function (that) {
        if (that instanceof TimesOperator) {
            var order = this._operand1.compareTo(that._operand1);
            if (order === 0) {
                order = this._operand2.compareTo(that._operand2);
            }
            return order;
        }
        return Objects.compare(this.typeOrder(), that.typeOrder());
    };
    TimesOperator.prototype.equals = function (that) {
        if (this === that) {
            return true;
        }
        else if (that instanceof TimesOperator) {
            return this._operand1.equals(that._operand1) && this._operand2.equals(that._operand2);
        }
        return false;
    };
    TimesOperator.prototype.hashCode = function () {
        if (TimesOperator._hashSeed === void 0) {
            TimesOperator._hashSeed = Murmur3.seed(TimesOperator);
        }
        return Murmur3.mash(Murmur3.mix(Murmur3.mix(TimesOperator._hashSeed, this._operand1.hashCode()), this._operand2.hashCode()));
    };
    TimesOperator.prototype.debug = function (output) {
        output.debug(this._operand1).write(46).write("times").write(40)
            .debug(this._operand2).write(41);
    };
    TimesOperator.prototype.clone = function () {
        return new TimesOperator(this._operand1.clone(), this._operand2.clone());
    };
    return TimesOperator;
}(BinaryOperator));
export { TimesOperator };
Item.TimesOperator = TimesOperator;
