// @swim/util/lib/main/CursorEmpty.js
import { __extends } from "tslib";
import { Cursor } from "./Cursor";
var CursorEmpty = (function (_super) {
    __extends(CursorEmpty, _super);
    function CursorEmpty() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CursorEmpty.prototype.isEmpty = function () {
        return true;
    };
    CursorEmpty.prototype.head = function () {
        throw new Error("empty");
    };
    CursorEmpty.prototype.step = function () {
        throw new Error("empty");
    };
    CursorEmpty.prototype.skip = function (count) {
    };
    CursorEmpty.prototype.hasNext = function () {
        return false;
    };
    CursorEmpty.prototype.nextIndex = function () {
        return 0;
    };
    CursorEmpty.prototype.next = function () {
        return { done: true };
    };
    CursorEmpty.prototype.hasPrevious = function () {
        return false;
    };
    CursorEmpty.prototype.previousIndex = function () {
        return -1;
    };
    CursorEmpty.prototype.previous = function () {
        return { done: true };
    };
    return CursorEmpty;
}(Cursor));
export { CursorEmpty };
Cursor.Empty = CursorEmpty;
